File: currency-rn/test/wire.test.ts
/**
 * Runs the frozen conformance vectors from spec/vectors.json.
 * Both projects must pass all 9 — see spec/wire.md.
 *
 *   node --experimental-strip-types --test currency-rn/test/wire.test.ts
 */

import test from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';

import {
  encode,
  decode,
  toB64url,
  fromB64url,
  wrapRelay,
  parseRelay,
  dedupeKey,
  isBearer,
  WireError,
  NOTE_LEN,
  type Note,
} from '../src/wire.ts';

const here = dirname(fileURLToPath(import.meta.url));
// Local fixture, copied from spec/vectors.json. Copied, never linked — no build
// artefact of this project may reach outside it (spec/wire.md, segregation note).
// Re-copy after any spec change; the freeze makes that a rare event.
const vectors = JSON.parse(readFileSync(resolve(here, 'vectors.json'), 'utf8'));

function bytesToHex(b: Uint8Array): string {
  return [...b].map((x) => x.toString(16).padStart(2, '0')).join('');
}

test('encode vectors', () => {
  for (const c of vectors.encode) {
    const raw = encode(c.note as Note);
    assert.equal(raw.length, NOTE_LEN, `${c.name}: length`);
    assert.equal(bytesToHex(raw), c.expect_hex, `${c.name}: hex`);
    assert.equal(toB64url(raw), c.expect_b64url, `${c.name}: b64url`);
    assert.deepEqual(decode(raw), c.note, `${c.name}: round-trip`);
    assert.deepEqual(
      decode(fromB64url(c.expect_b64url)),
      c.note,
      `${c.name}: b64url round-trip`,
    );
  }
});

test('decode rejections', () => {
  for (const c of vectors.decode) {
    assert.throws(
      () => decode(fromB64url(c.input_b64url)),
      (e: unknown) =>
        e instanceof WireError && e.code === c.expect_error,
      `${c.name}: expected ${c.expect_error}`,
    );
  }
});

test('base64url is 56 chars and unpadded', () => {
  for (const c of vectors.encode) {
    assert.equal(c.expect_b64url.length, 56, c.name);
    assert.ok(!c.expect_b64url.includes('='), c.name);
  }
});

test('bearer flag', () => {
  const bearer = vectors.encode.find((c: any) => c.name === 'bearer-holder-zero');
  assert.ok(isBearer(bearer.note));
  assert.equal(bearer.note.holder, '0'.repeat(32));
});

test('dedupe key is note_id + lamport', () => {
  const a = { ...vectors.encode[0].note, lamport: 1 } as Note;
  const b = { ...vectors.encode[0].note, lamport: 2 } as Note;
  assert.notEqual(dedupeKey(a), dedupeKey(b));
  assert.equal(dedupeKey(a), dedupeKey({ ...a }));
});

test('relay envelope round-trips and decrements', () => {
  const b64 = vectors.encode[0].expect_b64url;
  const env = wrapRelay(b64);
  assert.equal(env, `R|3|${b64}`);

  const parsed = parseRelay(env)!;
  assert.equal(parsed.ttl, 3);
  assert.equal(parsed.b64, b64);

  assert.equal(parseRelay(b64), null, 'direct transfer is not a relay envelope');

  const hop = wrapRelay(parsed.b64, parsed.ttl - 1);
  assert.equal(parseRelay(hop)!.ttl, 2);
});

File: currency-flutter/test/wire_test.dart
/// Runs the frozen conformance vectors. Both projects must pass all 9.
/// Contract: ../../spec/wire.md
///
///   cd currency-flutter && dart test
///
/// test/vectors.json is a COPY of spec/vectors.json — copied, never linked,
/// so no build artefact of this project reaches outside it.
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:test/test.dart';

import '../lib/wire.dart';

Note _noteFrom(Map<String, dynamic> m) => Note(
      ver: m['ver'] as int,
      flags: m['flags'] as int,
      noteId: m['note_id'] as String,
      value: m['value'] as int,
      holder: m['holder'] as String,
      lamport: m['lamport'] as int,
    );

String _hex(Uint8List b) =>
    b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();

void main() {
  final vectors = jsonDecode(
    File('test/vectors.json').readAsStringSync(),
  ) as Map<String, dynamic>;

  final encodeCases = (vectors['encode'] as List).cast<Map<String, dynamic>>();
  final decodeCases = (vectors['decode'] as List).cast<Map<String, dynamic>>();

  test('encode vectors', () {
    for (final c in encodeCases) {
      final name = c['name'];
      final note = _noteFrom(c['note'] as Map<String, dynamic>);
      final raw = encode(note);

      expect(raw.length, kNoteLen, reason: '$name: length');
      expect(_hex(raw), c['expect_hex'], reason: '$name: hex');
      expect(toB64url(raw), c['expect_b64url'], reason: '$name: b64url');
      expect(decode(raw), note, reason: '$name: round-trip');
      expect(decode(fromB64url(c['expect_b64url'] as String)), note,
          reason: '$name: b64url round-trip');
    }
  });

  test('decode rejections', () {
    for (final c in decodeCases) {
      expect(
        () => decode(fromB64url(c['input_b64url'] as String)),
        throwsA(isA<WireError>()
            .having((e) => e.code, 'code', c['expect_error'])),
        reason: c['name'] as String,
      );
    }
  });

  test('base64url is 56 chars and unpadded', () {
    for (final c in encodeCases) {
      expect((c['expect_b64url'] as String).length, 56, reason: c['name']);
      expect((c['expect_b64url'] as String).contains('='), isFalse,
          reason: c['name']);
    }
  });

  test('bearer flag', () {
    final c = encodeCases.firstWhere((c) => c['name'] == 'bearer-holder-zero');
    final note = _noteFrom(c['note'] as Map<String, dynamic>);
    expect(note.isBearer, isTrue);
    expect(note.holder, '0' * 32);
  });

  test('dedupe key is noteId + lamport', () {
    final base = _noteFrom(encodeCases[0]['note'] as Map<String, dynamic>);
    final a = Note(
        ver: base.ver,
        flags: base.flags,
        noteId: base.noteId,
        value: base.value,
        holder: base.holder,
        lamport: 1);
    final b = Note(
        ver: base.ver,
        flags: base.flags,
        noteId: base.noteId,
        value: base.value,
        holder: base.holder,
        lamport: 2);
    expect(a.dedupeKey, isNot(b.dedupeKey));
  });

  test('relay envelope round-trips and decrements', () {
    final b64 = encodeCases[0]['expect_b64url'] as String;
    final env = wrapRelay(b64);
    expect(env, 'R|3|$b64');

    final parsed = parseRelay(env)!;
    expect(parsed.ttl, 3);
    expect(parsed.b64, b64);

    expect(parseRelay(b64), isNull,
        reason: 'direct transfer is not a relay envelope');

    final hop = wrapRelay(parsed.b64, parsed.ttl - 1);
    expect(parseRelay(hop)!.ttl, 2);
  });
}
